### PR TITLE
TMA 712 add params checking against the specification in all actions

### PR DESCRIPTION
**Pull request by Hahihula**

### CODE REVIEW

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 1]
```diff
@@ -63,7 +137,10 @@ def call(params)
           symbolized_config[:labels] = symbolized_config[:labels].map { |l| GoodData::Helpers.symbolize_keys(l) }
           headers_in_options = params.csv_headers == 'false' || true
 
-          mode = params.sync_mode || 'sync_project'
+          mode = params.sync_mode
+          if mode.nil? || MODES.include?(mode).!
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Performs a nil-check (NilCheck)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 2]
```diff
@@ -42,16 +42,90 @@ class SynchronizeUserFilters < BaseAction
...
+          sync_domain_client_workspaces
+        )
+
         def call(params)
           client = params.gdc_gd_client
           domain_name = params.organization || params.domain
+          fail "Either organisation or domain has to be specified in params" unless domain_name
           domain = client.domain(domain_name) if domain_name
           project = client.projects(params.gdc_project) || client.projects(params.gdc_project_id)
+          fail "Either project or project_id has to be specified in params" unless project
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Tests 'project' at least 4 times (RepeatedConditional)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 3]
```diff
@@ -48,11 +101,13 @@ def version
...
           client = params.gdc_gd_client
           domain_name = params.organization || params.domain
+          fail "Either organisation or domain has to be specified in params" unless domain_name
           project = client.projects(params.gdc_project) || client.projects(params.gdc_project_id)
+          fail "Either project or project_id has to be specified in params" unless project
           data_source = GoodData::Helpers::DataSource.new(params.input_source)
           data_product = params.data_product
           mode = params.sync_mode
-          unless mode.nil? || MODES.include?(mode)
+          if mode.nil? || MODES.include?(mode).!
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Performs a nil-check (NilCheck)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 4]
```diff
@@ -48,11 +101,13 @@ def version
         def call(params)
           client = params.gdc_gd_client
           domain_name = params.organization || params.domain
+          fail "Either organisation or domain has to be specified in params" unless domain_name
           project = client.projects(params.gdc_project) || client.projects(params.gdc_project_id)
+          fail "Either project or project_id has to be specified in params" unless project
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Tests 'project' at least 4 times (RepeatedConditional)

File: lib/gooddata/lcm/helpers/check_helper.rb [hunk 5]
```diff
@@ -14,9 +14,10 @@ def check_params(specification, params)
           specification.keys.each do |param_name|
             value = params.send(param_name)
             type = specification[param_name][:type]
-
-            if value.nil?
-              if specification[param_name][:opts][:required]
+            if value.nil? || ((value.is_a? String) && value.empty?)
+              if specification[param_name][:opts][:default]
```
Comment (yenkins) [inline lib/gooddata/lcm/helpers/check_helper.rb]: Calls 'specification[param_name]' 6 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/helpers/check_helper.rb]: Calls 'specification[param_name][:opts]' 5 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/helpers/check_helper.rb]: Calls 'specification[param_name][:opts][:default]' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/helpers/check_helper.rb]: Refers to 'specification' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/lcm/helpers/check_helper.rb [hunk 6]
```diff
@@ -30,7 +31,7 @@ def check_params(specification, params)
               end
 
               unless type.check(value)
-                fail "Parameter '#{param_name}' has invalid type, expected: #{type}"
+                fail "Parameter '#{param_name}' has invalid type, expected: #{type}, got #{value.class}"
```
Comment (yenkins) [inline lib/gooddata/lcm/helpers/check_helper.rb]: Calls 'value.class' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/helpers/check_helper.rb [hunk 7]
```diff
@@ -14,9 +14,10 @@ def check_params(specification, params)
           specification.keys.each do |param_name|
             value = params.send(param_name)
             type = specification[param_name][:type]
-
-            if value.nil?
-              if specification[param_name][:opts][:required]
+            if value.nil? || ((value.is_a? String) && value.empty?)
```
Comment (yenkins) [inline lib/gooddata/lcm/helpers/check_helper.rb]: Performs a nil-check (NilCheck)

File: lib/gooddata/lcm/lcm2.rb [hunk 8]
```diff
@@ -311,9 +325,11 @@ def perform(mode, params = {})
             GoodData.logger.info("Running #{action.name} action ...")
 
             # Check if all required parameters were passed
+            params.clear_filters
             BaseAction.check_params(action.const_get('PARAMS'), params)
-
+            params.set_filters(action.const_get('PARAMS'))
```
Comment (yenkins) [inline lib/gooddata/lcm/lcm2.rb]: Calls 'action.const_get('params')' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/lcm2.rb [hunk 9]
```diff
@@ -311,9 +325,11 @@ def perform(mode, params = {})
             GoodData.logger.info("Running #{action.name} action ...")
 
             # Check if all required parameters were passed
+            params.clear_filters
```
Comment (yenkins) [inline lib/gooddata/lcm/lcm2.rb]: Calls 'params.clear_filters' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/lcm2.rb]: Refers to 'params' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/lcm/types/class/gd_logger.rb [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+# LICENSE file in the root directory of this source tree.
+
+require_relative '../base_type'
+
+require_relative 'class'
+
+module GoodData
+  module LCM2
+    module Type
+      class GdLogger < ClassType
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/gd_logger.rb]: Has no descriptive comment (IrresponsibleModule)

File: lib/gooddata/lcm/types/class/gd_logger.rb [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+# Copyright (c) 2010-2017 GoodData Corporation. All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+require_relative '../base_type'
+
+require_relative 'class'
+
+module GoodData
+  module LCM2
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/gd_logger.rb]: Has the name 'lcm2' (UncommunicativeModuleName)

File: lib/gooddata/lcm/types/class/gd_logger.rb [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    module Type
+      class GdLogger < ClassType
+        CATEGORY = :class
+
+        def check(value)
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/gd_logger.rb]: Doesn't depend on instance state (maybe move it to another class?) (UtilityFunction)

File: lib/gooddata/lcm/types/class/gd_product.rb [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+# LICENSE file in the root directory of this source tree.
+
+require_relative '../base_type'
+
+require_relative 'class'
+
+module GoodData
+  module LCM2
+    module Type
+      class GdProductType < ClassType
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/gd_product.rb]: Has no descriptive comment (IrresponsibleModule)

File: lib/gooddata/lcm/types/class/gd_product.rb [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+# Copyright (c) 2010-2017 GoodData Corporation. All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+require_relative '../base_type'
+
+require_relative 'class'
+
+module GoodData
+  module LCM2
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/gd_product.rb]: Has the name 'lcm2' (UncommunicativeModuleName)

File: lib/gooddata/lcm/types/class/gd_product.rb [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+    module Type
+      class GdProductType < ClassType
+        CATEGORY = :class
+
+        def check(value)
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/gd_product.rb]: Doesn't depend on instance state (maybe move it to another class?) (UtilityFunction)

File: lib/gooddata/lcm/types/class/gd_project.rb [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+# LICENSE file in the root directory of this source tree.
+
+require_relative '../base_type'
+
+require_relative 'class'
+
+module GoodData
+  module LCM2
+    module Type
+      class GdProjectType < ClassType
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/gd_project.rb]: Has no descriptive comment (IrresponsibleModule)

File: lib/gooddata/lcm/types/class/gd_project.rb [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+# Copyright (c) 2010-2017 GoodData Corporation. All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+require_relative '../base_type'
+
+require_relative 'class'
+
+module GoodData
+  module LCM2
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/gd_project.rb]: Has the name 'lcm2' (UncommunicativeModuleName)

File: lib/gooddata/lcm/types/class/gd_project.rb [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+    module Type
+      class GdProjectType < ClassType
+        CATEGORY = :class
+
+        def check(value)
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/gd_project.rb]: Doesn't depend on instance state (maybe move it to another class?) (UtilityFunction)

File: lib/gooddata/lcm/types/class/smart_hash.rb [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+# LICENSE file in the root directory of this source tree.
+
+require_relative '../base_type'
+
+require_relative 'class'
+
+module GoodData
+  module LCM2
+    module Type
+      class GdSmartHashType < ClassType
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/smart_hash.rb]: Has no descriptive comment (IrresponsibleModule)

File: lib/gooddata/lcm/types/class/smart_hash.rb [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+# Copyright (c) 2010-2017 GoodData Corporation. All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+require_relative '../base_type'
+
+require_relative 'class'
+
+module GoodData
+  module LCM2
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/smart_hash.rb]: Has the name 'lcm2' (UncommunicativeModuleName)

File: lib/gooddata/lcm/types/class/smart_hash.rb [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+        #   param :username, instance_of(Type::StringType), required: true
+        #
+        #   description 'Password used for connecting to GD'
+        #   param :password, instance_of(Type::StringType), required: true
+        #
+        #   description 'Hostname'
+        #   param :hostname, instance_of(Type::StringType), required: false, default: 'secure.gooddata.com'
+        # end
+
+        def check(value)
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/smart_hash.rb]: Doesn't depend on instance state (maybe move it to another class?) (UtilityFunction)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 22]
```diff
@@ -25,6 +25,59 @@ class SynchronizeUsers < BaseAction
...
+
+        description 'GDC Project'
+        param :gdc_project, instance_of(Type::GdProjectType), required: false
+
+        description 'GDC Project Id'
+        param :gdc_project_id, instance_of(Type::StringType), required: false
+
+        description 'Fail Early'
+        param :fail_early, instance_of(Type::BooleanType), required: false
+
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Style/TrailingWhitespace: Trailing whitespace detected.

File: lib/gooddata/lcm/actions/update_release_table.rb [hunk 23]
```diff
@@ -14,6 +14,25 @@ class UpdateReleaseTable < BaseAction
...
+
+        description 'Segments to manage'
+        param :segments, array_of(instance_of(Type::SegmentType)), required: true
+
+        description 'Table Name'
+        param :release_table_name, instance_of(Type::StringType), required: false
+
+        description 'Query'
+        param :query, instance_of(Type::GdSmartHashType), required: false
+
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/update_release_table.rb]: Style/EmptyLinesAroundBlockBody: Extra empty line detected at block body end.

File: lib/gooddata/lcm/lcm2.rb [hunk 24]
```diff
@@ -35,6 +36,19 @@ def method_missing(name, *_args)
...
+      def [](variable)
+        fail "Param #{variable} is not defined in the specification" if @specification && !@specification[variable] && !@specification[variable.to_sym]
+        super(variable)
+      end
+
+      def clear_filters
+        @specification = nil
+      end
+
+      def set_filters(filter)
```
Comment (yenkins) [inline lib/gooddata/lcm/lcm2.rb]: Style/AccessorMethodName: Do not prefix writer method names with `set_`.

File: spec/unit/helpers/check_helper_spec.rb [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+      description 'Testing param four'
+      param :test_param_four, instance_of(GoodData::LCM2::Type::StringType), required: false
+    end
+    expect { GoodData::LCM2::Helpers.check_params(PARAMS_3, params) }.to raise_error(/has invalid type/)
+  end
+
+  it 'fails when unspecified variable is acessed' do
+    GoodData::LCM2::Helpers.check_params(PARAMS_2, params)
+    params.set_filters(PARAMS_2)
+    expect{params[:test_param_three]}.to raise_error(/not defined in the specification/)
```
Comment (yenkins) [inline spec/unit/helpers/check_helper_spec.rb]: Style/SpaceBeforeBlockBraces: Space missing to the left of {.

Comment (yenkins) [inline spec/unit/helpers/check_helper_spec.rb]: Style/SpaceInsideBlockBraces: Space missing inside {.

Comment (yenkins) [inline spec/unit/helpers/check_helper_spec.rb]: Style/SpaceInsideBlockBraces: Space missing inside }.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/703/ : SUCCESS in 13s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/536/ : FAILURE in 2m 35s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/813/ : FAILURE in 1m 58s

File: spec/unit/lcm/lcm2_spec.rb [hunk 26]
```diff
@@ -9,13 +9,14 @@
...
-    let(:logger) { double(:logger) }
+    let(:domain) { 'domain' }
+    let(:logger) { GoodData.logger }
     let(:params) do
       params = {
         skip_actions: %w(CollectSegments SynchronizeUsers),
         GDC_GD_CLIENT: client,
-        GDC_LOGGER: logger
+        GDC_LOGGER: logger,
+        domain: domain,
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: Style/TrailingCommaInLiteral: Avoid comma after the last item of a hash.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/704/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/537/ : FAILURE in 46s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/814/ : SUCCESS in 34s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/705/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/538/ : FAILURE in 44s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/815/ : SUCCESS in 1m 56s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/706/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/539/ : SUCCESS in 2m 03s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/816/ : FAILURE in 2m 00s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/707/ : SUCCESS in 19s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/540/ : SUCCESS in 1m 56s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/817/ : SUCCESS in 1m 54s

**Pull request closed without merging**